test/integration/master/secrets_enveloping_test.go

```diff
@@ -0,0 +1,313 @@
+	"k8s.io/apiserver/pkg/storage/storagebackend"
+	"k8s.io/apiserver/pkg/storage/value"
+	aestransformer "k8s.io/apiserver/pkg/storage/value/encrypt/aes"
+	"k8s.io/client-go/kubernetes"
+	kubeapiservertesting "k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+const (
+	testNamespace = "secret-encryption-test"
+	testSecret    = "test-secret"
+
+	aesGCMPrefix = "k8s:enc:aesgcm:v1:key1:"
+	aesCBCPrefix = "k8s:enc:aescbc:v1:key1:"
+
+	// Secret Data
+	secretKey = "api_key"
+	secretVal = "086a7ffc-0225-11e8-ba89-0ed5f89f718b"
+
+	aesGCMConfigYAML = `
+kind: EncryptionConfig
+apiVersion: v1
+resources:
+  - resources:
+    - secrets
+    providers:
+    - aesgcm:
+        keys:
+        - name: key1
+          secret: c2VjcmV0IGlzIHNlY3VyZQ==
+`
+
+	aesCBCConfigYAML = `
+kind: EncryptionConfig
+apiVersion: v1
+resources:
+  - resources:
+    - secrets
+    providers:
+    - aescbc:
+        keys:
+        - name: key1
+          secret: c2VjcmV0IGlzIHNlY3VyZQ==
+`
+)
+
+type unSealSecret func(cipherText []byte, ctx value.Context, config encryptionconfig.ProviderConfig) ([]byte, error)
+
+var testCases = []struct {
+	transformerConfigContent string
+	transformerPrefix        string
+	unSealFunc               unSealSecret
+}{
+	{aesGCMConfigYAML, aesGCMPrefix, unSealWithGCMTransformer},
+	{aesCBCConfigYAML, aesCBCPrefix, unSealWithCBCTransformer},
+	// TODO: add secretbox
+}
+
+// TestSecretsShouldBeEnveloped is an integration test between KubeAPI and ECTD that checks:
+// 1. Secrets are encrypted on write
+// 2. Secrets are decrypted on read
+// when EncryptionConfig is passed to KubeAPI server.
+func TestSecretsShouldBeEnveloped(t *testing.T) {
+	for _, tt := range testCases {
+		runEnvelopeTest(t, tt.unSealFunc, tt.transformerConfigContent, tt.transformerPrefix)
+	}
+}
+
+func runEnvelopeTest(t *testing.T, unSealSecretFunc unSealSecret, transformerConfigYAML, expectedEnvelopePrefix string) {
+	transformerConfig := parseTransformerConfigOrDie(t, transformerConfigYAML)
+
+	storageConfig := framework.SharedEtcd()
+	kubeAPIServer := startKubeApiWithEncryptionOrDie(t, storageConfig, transformerConfigYAML)
+	defer kubeAPIServer.TearDownFn()
+
+	client, err := kubernetes.NewForConfig(kubeAPIServer.ClientConfig)
+	if err != nil {
+		t.Fatalf("error while creating client: %v", err)
+	}
+
+	ns, err := createTestNamespace(client, testNamespace)
+	if err != nil {
+		t.Error(err)
+	}
+	defer func() {
+		client.CoreV1().Namespaces().Delete(ns.Name, metav1.NewDeleteOptions(0))
+	}()
+
+	_, err = createTestSecret(client, testSecret, ns.Name)
+	if err != nil {
+		t.Error(err)
+	}
+
+	etcdPath := getETCDPath(storageConfig.Prefix)
+	response, err := readRawRecordFromETCD(kubeAPIServer, etcdPath)
+	if err != nil {
+		t.Error(err)
+	}
+
+	if !bytes.HasPrefix(response.Kvs[0].Value, []byte(expectedEnvelopePrefix)) {
+		t.Errorf("expected secrets to be enveloped by %s", expectedEnvelopePrefix)
```

Comment:
Could you improve the error message by also adding the actual prefix of the secret? This code simplify debugging.

Reply:
I modified the error message to print the whole secret (in addition to the expected prefix) since there may be cases where no envelope prefix is present.